bin/btc-server/src/database/mod.rs

```
@@ -329,6 +348,164 @@ impl Db {
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the key package is successfully stored.
+    /// Returns `Err` in case of serialization or other errors.
+    pub fn set_key_package_by_id(
+        &self,
+        multisig_id: u32,
+        key_package: frost::keys::KeyPackage,
+    ) -> Result<(), Error> {
+        let key = multisig_id.to_le_bytes();
+        let mut bytes = Vec::new();
+        ciborium::into_writer(&key_package, &mut bytes).expect("writing to buffer");
+
+        self.key_packages.insert(&key, &bytes[..])?;
+        Ok(())
+    }
+
+    /// Sets a public key package by multisig_id in the multi-key storage.
+    ///
+    /// # Arguments
+    ///
+    /// * `multisig_id` - The u32 identifier for the multisig/federation.
+    /// * `pk_package` - The `frost::keys::PublicKeyPackage` to be stored.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if the public key package is successfully stored.
+    /// Returns `Err` in case of serialization or other errors.
+    pub fn set_pubkey_package_by_id(
+        &self,
+        multisig_id: u32,
+        pk_package: frost::keys::PublicKeyPackage,
+    ) -> Result<(), Error> {
+        let key = multisig_id.to_le_bytes();
+        let mut bytes = Vec::new();
+        ciborium::into_writer(&pk_package, &mut bytes).expect("writing to buffer");
+
+        self.pubkey_packages.insert(&key, &bytes[..])?;
+        Ok(())
+    }
+
+    /// Lists all multisig_ids that have key packages stored.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Vec<u32>` containing all multisig_ids that have key packages.
+    /// Returns an empty vector if no key packages are found.
+    /// Returns `Err` in case of database errors.
+    pub fn list_multisig_ids(&self) -> Result<Vec<u32>, Error> {
+        let mut ids = Vec::new();
+        for res in self.key_packages.iter().keys() {
+            let key_bytes = res?;
+            if key_bytes.len() == 4 {
```

I'd return an error; this implies a corrupted DB.

---

good point!